enhancements/network/optional-on-prem-dns-records.md

```diff
@@ -0,0 +1,219 @@
+  the internal loadbalancer and DNS infrastructure is increasing all the time,
+  and at this point it is in use for more platforms than it isn't. Making these
+  options platform-independent would significantly reduce the duplication
+  across the different infrastructure objects.
+
+  While this may not be strictly required for this specific feature, if we're
+  going to move these options it should probably happen before we add more to
+  the existing location as that will just make it more difficult to change
+  later.
+
+Other options were considered, but rejected for the following reasons:
+
+* Add another type to the existing loadbalancer setting that would disable
+  both the loadbalancer and DNS services. This was not favored because the
+  DNS infrastructure is distinct from the loadbalancer and it would overload
+  the meaning of the loadbalancer setting. Additionally, it could get quite
+  messy if we add any more options to either the loadbalancer or DNS. We
+  would have to enumerate every possible combination of options. With only
+  two per service it wouldn't be too bad, but it could quickly escalate.
+
+* Support multiple VIPs in the coredns configuration. This would have
+  significantly increased the complexity of our coredns configuration
+  and would have introduced the possibility of variance between how DNS
+  resolves internally or externally. Additionally, multiple backends
+  behind a single DNS name is already supported by the external infrastructure
+  so at best we'd be duplicating existing functionality.
+
+* Remove coredns completely. This has a few problems. As noted earlier, it
+  would require changing the resolv-prepender service to handle the absence
+  of a local DNS server, and that's a much riskier change. It would also
+  eliminate name-based resolution between nodes.
+
+## Open Questions [optional]
+
+NA
+
+## Test Plan
+
+We will need to add the ability to configure api-int on the external DNS in
+our external loadbalancer tests. Since we already have an api record, that
+shouldn't be difficult.
+
+## Graduation Criteria
+
+We have a specific customer waiting for this feature who has agreed to test
+it in tech preview. Graduation will be predicated on the feature addressing
+their use case.
+
+### Dev Preview -> Tech Preview
+
+NA
+
+### Tech Preview -> GA
+
+Once the customer has verified that this addresses their use case it should be
+ready for GA.
+
+### Removing a deprecated feature
+
+NA
+
+## Upgrade / Downgrade Strategy
+
+Currently this can only be set at install time, so it is not relevant for
+upgrades and downgrades
+
+## Version Skew Strategy
+
+There will also be no version skew.
+
+## Operational Aspects of API Extensions
+
+Because they can only be set at deploy-time I don't expect a lot of impact here.
+When debugging problems we'll need to look at the Infrastructure record to see
+whether internal DNS records are in use.
```

Comment:
Is there a particular reason this has to be install-time only?

Is it possible that users with internal DNS records configured today but using an external LB may want to transition their cluster to not having internal DNS records configured? Should that be something we support?

Reply:
Most of our configuration can only be set at install-time. There are a couple of exceptions where we found things that needed to be (and could be) modified on day 2, but in general making infrastructure-level changes after initial deployment is tricky and adds yet another path we need to test.

This one probably could be modified on day 2 without too much difficulty, but right now I'm also limiting it to when a user-managed loadbalancer is in use, and since we don't allow changing that on day 2 I don't think we need to allow this either. If we decide to start supporting that migration we can revisit this.